src/lib/gadgets/bitwise.ts

```diff
@@ -10,7 +10,44 @@ import {
   divideWithRemainder,
 } from './common.js';
 
-export { xor, rotate };
+export { xor, not, rotate };
+
+function not(a: Field, length: number) {
+  // check that input length is positive
+  assert(length > 0, `Input length needs to be positive values.`);
+
+  // Check that length does not exceed maximum field size in bits
+  assert(
+    length <= Field.sizeInBits(),
+    `Length ${length} exceeds maximum of ${Field.sizeInBits()} bits.`
+  );
+
+  // obtain pad length until the length is a multiple of 16 for n-bit length lookup table
+  let padLength = Math.ceil(length / 16) * 16;
+
+  // handle constant case
+  if (a.isConstant()) {
+    let max = 1n << BigInt(padLength);
+    assert(
+      a.toBigInt() < max,
+      `${a.toBigInt()} does not fit into ${padLength} bits`
+    );
+    return new Field(Fp.not(a.toBigInt()));
+  }
+
+  // create a bitmask with all ones
+  let allOnesF = new Field(2n ** BigInt(length) - 1n);
+
+  let allOnes = Provable.witness(Field, () => {
+    return allOnesF;
+  });
+
+  allOnesF.assertEquals(allOnes);
+
+  let notOutput = xor(a, allOnes, length);
```

Comment:
I think we should actually implement the unchecked version `bnot_unchecked`, it saves a good amount of constraints (`bnot_unchecked` has only 1 constraint, no matter how large the bit length gets whereas `bnot_checked` increases in constraints with the length of the bit size because it generates a bunch of XOR gates). 

This is actually only a simple change - instead of calling into `xor`, instead its a simple subtraction

```ocaml
  (* Negating is equivalent to subtracting with all one word *)
  (* [2^len - 1] - input = not (input) *)
  Generic.sub (module Circuit) all_ones_var input
```

This could also be a conditional check wdyt @mitschabaude @MartinMinkov

Reply:
Just to clarify for myself, why wouldn't someone want to use the version with a checked length? Is the idea that the length is checked elsewhere and then does not have to be redone in the NOT gate? If that's the case, then we are trading off API safety/simplicity with low-level power usage. This seems fine to me! I would prefer adding a conditional check/flag over adding another unchecked version.

Reply:
I slightly prefer adding a flag over another unchecked version. The api could be something like this.
```ts
not(a: Field, length: number, checked: boolean = true)
```
We can set the default to checked or even unchecked.

Reply:
@MartinMinkov we have many "data types with intrinsic constraints", like `Bool` or `UInt64` which always get constraints attached when introduced in a circuit. We _never_ additionally check those constraints in operations. For example, `Bool.and()` doesn't check that its inputs are booleans, `UInt64.add()` doesn't check that its inputs fit in 64 bits and so on. The rule is that you should constraint outputs, but not inputs. IMO doing the checked version by default would violate that rule.

@ymekuria I therefore think the _unchecked_ version should be the default. I'm in favor of the flag if we can find any case where the checked version is useful, i.e. better than just doing `rangeCheck64(x); not(x)`.

@querolita maybe you know why we had two versions here? Is there a case where the checked version is better?